### PR TITLE
#1059: Add support for floating numbers separator

### PR DESCRIPTION
**Pull request by koluch**
#1059: Add support for floating numbers separator

### CODE REVIEW

**Comment by MaxGraey** [comment]
Thanks! You should add your name and email to NOTICE file first.

**Comment by MaxGraey** [comment]
It also makes sense to handle and throw error for invalid cases like:
```ts
_123.0
123_.0
123._0
123.0_
0_0.0_0
```

File: src/tokenizer.ts [hunk 1]
```diff
@@ -1522,12 +1527,14 @@ export class Tokenizer extends DiagnosticEmitter {
...
-        while (this.pos < end && isDecimalDigit(text.charCodeAt(this.pos))) {
+        while (this.pos < end) {
+          let ch = text.charCodeAt(this.pos);
+          if (!(isDecimalDigit(ch) || isDecimalSeparator(ch))) break;
           ++this.pos;
         }
       }
     }
-    return parseFloat(text.substring(start, this.pos));
+    return parseFloat(text.substring(start, this.pos).replace(/_/g, ''));
```
Comment (MaxGraey) [inline src/tokenizer.ts]: Also we try to avoid RegExp

File: src/util/charcode.ts [hunk 2]
```diff
@@ -174,7 +174,11 @@ export function isWhiteSpace(c: i32): bool {
 
 /** Tests if the specified character code is a valid decimal digit. */
 export function isDecimalDigit(c: i32): bool {
-  return c >= CharCode._0 && c <= CharCode._9;
+  return c >= CharCode._0 && c <= CharCode._9 ;
```
Comment (MaxGraey) [inline src/util/charcode.ts]: nit: unnecessary space between semicolon and `_9`

File: src/tokenizer.ts [hunk 3]
```diff
@@ -1499,16 +1500,20 @@ export class Tokenizer extends DiagnosticEmitter {
...
   readDecimalFloat(): f64 {
-    // TODO: numeric separators (parseFloat can't handle these)
     var start = this.pos;
     var end = this.end;
     var text = this.source.text;
-    while (this.pos < end && isDecimalDigit(text.charCodeAt(this.pos))) {
+
+    while (this.pos < end) {
+      let ch = text.charCodeAt(this.pos);
+      if (!(isDecimalDigit(ch) || isDecimalSeparator(ch))) break;
```
Comment (dcodeIO) [inline src/tokenizer.ts]: This looks like it would accept `___123___123___.___123___`. Is that valid?

Comment (dcodeIO) [reply src/tokenizer.ts:1509]: Any thoughts on this? :)

**Comment by dcodeIO** [comment]
[2020 vacuum](https://github.com/AssemblyScript/assemblyscript/issues/1277) as it appears to be stale. Looks like not much was missing here to get it working, so feel free to open a new PR addressing the comments left so far.

**Pull request closed without merging**